haystack/tools/pipeline_tool.py

```diff
@@ -0,0 +1,243 @@
+        name="document_retriever",
+        description="Retrieve documents relevant to a query from the document store",
+    )
+
+    # Create an Agent with the tool
+    agent = Agent(
+        chat_generator=OpenAIChatGenerator(model="gpt-4.1-mini"),
+        tools=[retriever_tool]
+    )
+
+    # Let the Agent handle a query
+    result = agent.run([ChatMessage.from_user("Who was Nikola Tesla?")])
+
+    # Print result of the tool call
+    print("Tool Call Result:")
+    print(result["messages"][2].tool_call_result.result)
+    print("")
+
+    # Print answer
+    print("Answer:")
+    print(result["messages"][-1].text)
+    ```
+    """
+
+    def __init__(
+        self,
+        pipeline: Union[Pipeline, AsyncPipeline],
+        *,
+        name: str,
+        description: str,
+        input_mapping: Optional[dict[str, list[str]]] = None,
+        output_mapping: Optional[dict[str, str]] = None,
+        parameters: Optional[dict[str, Any]] = None,
+        outputs_to_string: Optional[dict[str, Union[str, Callable[[Any], str]]]] = None,
+        inputs_from_state: Optional[dict[str, str]] = None,
+        outputs_to_state: Optional[dict[str, dict[str, Union[str, Callable]]]] = None,
+    ) -> None:
+        """
+        Create a Tool instance from a Haystack pipeline.
+
+        :param pipeline: The Haystack pipeline to wrap as a tool.
+        :param name: Name of the tool.
+        :param description: Description of the tool.
+        :param input_mapping: A dictionary mapping component input names to pipeline input socket paths.
+            If not provided, a default input mapping will be created based on all pipeline inputs.
+            Example:
+            ```python
+            input_mapping={
+                "query": ["retriever.query", "prompt_builder.query"],
+            }
+            ```
+        :param output_mapping: A dictionary mapping pipeline output socket paths to component output names.
+            If not provided, a default output mapping will be created based on all pipeline outputs.
+            Example:
+            ```python
+            output_mapping={
+                "retriever.documents": "documents",
+                "generator.replies": "replies",
+            }
+            ```
+        :param parameters:
+            A JSON schema defining the parameters expected by the Tool.
+            Will fall back to the parameters defined in the component's run method signature if not provided.
+        :param outputs_to_string:
+            Optional dictionary defining how a tool outputs should be converted into a string.
+            If the source is provided only the specified output key is sent to the handler.
+            If the source is omitted the whole tool result is sent to the handler.
+            Example:
+            ```python
+            {
+                "source": "docs", "handler": format_documents
+            }
+            ```
+        :param inputs_from_state:
```

Comment:
Did we add that document about states and tools? If so, it would be nice to reference it here

Reply:
Not yet issue is still open here https://github.com/deepset-ai/haystack/issues/9738